### ISSUE TITLE
A suggestion about managing dependencies

### ISSUE THREAD

**Issue by mkaze** [RasaHQ/whatlies#88]
Currently, I am working on the tf-hub and huggingface integrations with whatlies. However, there is something which is annoying me for a while which I'd like to discuss here and see if it's really an issue, or it's just me overthinking too much!

### Problem
As this library grows, more and more language/framework backends are supported, which is really good; however, this in turn has the side-effect of adding more and more dependency packages which provide the backbone of supporting a new language/framework. This might create two issues:

1. The chance of dependency conflict increases.
2. And more importantly, with current structure if a user just want to quickly try the base features of this library on their own embeddings (produced from their own custom model) or even work with only one of the minimal languages (say fasttext), they are forced to download and install all of these dependencies; and some of the dependencies, e.g. tensorflow and pytorch for hugging-face, are considerably large and itself have lots of additional dependencies. This may (or may not!) be frustrating, especially for a newbie user (e.g. in the past five years which I have worked with various versions of keras as well as tensorflow from 0.x to 1.x and then 2.x versions, I have had the pleasure of dealing with downloading and installing increasingly large wheel files and doing additional configurations, especially for GPU support; although, recently the additional configs have been considerably reduced/made easier).

### Solution
One approach is to separate each of these language/framework backends, and let the user to decide what they really need. For example, the following guide could be provided as installation instruction in readme file:

<hr/>

If you want to use the base features of this library (without the support for language backends), you can simply run:
```
pip install whatlies
```
However, if you want to install full feature set, including the support for all the language backends, you can use the following command:
```
pip install whatlies[full]
```
Alternatively, if you need the support for only one or some of the language backends and want to be selective about that (probably, due to not downloading and installing a lot of unnecessary dependencies packages), you can simply provide the name of backend(s) in square bracket, using the following table as a guide:

Language backend | Installation name
--------------------------|----------------------------
Fasttext | `fasttext`
Gensim | `gensim`
HuggingFace | `huggingface`
Spacy | `spacy`

For example, the installation commands in this case would look like as follow:
```
pip install whatlies[fasttext]         # install base features + Fasttext language backend
pip install whatlies[fasttext,gensim]  # install base features + Fasttext and Gensim language backends
```
<hr/>

As for the modifications to source code so that no import fails, two options comes to my mind: either we can defer the `import`s to be done in methods/functions:

```python
from whatlies import EmebddingSet


class DummyLanguage():
    def __init__(model=None):
        import Dummy
        ...
```

or, alternatively we can use `try/except` block to prevent errors about missing dependencies:

```python
from whatlies import EmebddingSet
try:
    import Dummy
except ImportError:
    pass


class DummyLanguage():
    ...
```

As you may already know, one example of managing dependencies with this approach is Rasa ([1](1), [2](2)).

Finally, I want to emphasis again that this might not be really an issue (or at least it's not an issue at this stage of development of this library) and it's likely that I am overthinking it too much. Anyways, I would really like to hear your thoughts on this, @koaning !

[1]: https://github.com/RasaHQ/rasa/blob/b70e9f4bbeff16fa9a39093d7a7d41dae9445e3e/rasa/nlu/utils/spacy_utils.py#L45
[2]: https://rasa.com/docs/rasa/user-guide/installation/#nlu-pipeline-dependencies

**Comment by koaning** [comment]
This is making me grin, basically because this issue is exactly the reason why I've avoided these two features. 

I wanted to post this immediately. I'll now read your suggestings and give it a proper ponder.

**Comment by koaning** [comment]
There's another package that I maintain that is [starting to also experience this issue](https://github.com/koaning/scikit-lego/issues/380). You raise an interesting point in terms of downloading language backends. This could be made easier by also adding a command line interface. I've made this a seperate issue [here](https://github.com/RasaHQ/whatlies/issues/90) so this thread can stick to discussing dependencies.

**Comment by koaning** [comment]
My main concern at the moment is that we might get into a sticky situation with items from huggingface and tfhub. The moment that a supported model from `tf-hub` requires a different version than a model from `huggingface` effectively we'd be stuck in the mud. 

I'm not too worried about spaCy/gensim as these backends are well maintained (I recall gensim even being a dependency of spaCy). I'm also not worried about fasttext because that package rarely gets an update. It's more the tf-hub and huggingface libraries that concern me here.

**Comment by mkaze** [comment]
> You raise an interesting point in terms of downloading language backends. This could be made easier by also adding a command line interface.

Yeah, that might be a good idea to make the process of downloading language files easier for the user.

> This is making me grin, basically because this issue is exactly the reason why I've avoided these two features.
...
> The moment that a supported model from tf-hub requires a different version of tensorflow than a model from huggingface effectively we'd be stuck in the mud.

Unlike you, I am not worried about this specific issue because I think it wouldn't happen. You should consider that TF 2.x is much more mature now than, say, one year ago. And it would be the future of TF because there is so much development investment poured into it. Besides, huggingface relies on TF 2 and also most of the models in tf-hub works with TF 2 now (and when I say TF 2 here, I mean they are compatible with all the TF 2.x versions). So, as I said, I am not worried about this.

In terms of whether this approach is an established way of managing optional (or hard vs. soft) dependencies, I would say yes; since some of the popular packages are using this approach (though, each of them with small differences to each other in terms of handling `import` statements).

However whether this should be done for `whatlies` as well, I would say I am not sure yet. As I mentioned previously, my main concern is not dependency conflict (although, it might happen especially when you have so much different dependencies where most of them is actively developed, and also considering the fact that even stable packages such as spacy or spacy-transformes are going under a complete backend overhaul now); rather, I am mostly concerned with the fact that for just starting to use this library, the user has to download and install around 2 GB of packages first. Although, one might point out that some of the language files (for example, for fasttext) or model file are much larger and therefore in comparison it's not an issue. And that's why I am not sure if this is a good move, or we should continue as before.

I think, at the end besides the possible technical issues, it mostly depends on the target user group of this library and how diverse their needs are when using this library; whether they are mostly experienced users in the field and programming, or they are mostly newcomers to the field or researchers. And, at the same time, I also believe that the usage of libraries/packages should be made simple for the users, but not simpler (thanks to Einstein! :D ).

**Comment by koaning** [comment]
I actually don't know who the users are. It has 15K downloads after 3 months ... so folks are using it. But they could be students, researches or industry. Dunno. 

I agree with your Einstein quote. My variant: keep the complex stuff complex in order to keep the simple things simple. 

I'll keep this thread around but I doubt I'll invest in the short term. Once the packages becomes a bottleneck it'll move up the priority list.

**Comment by koaning** [comment]
Gah. Tensorflow errors though. 

```
---------------------------------------------------------------------------
NotFoundError                             Traceback (most recent call last)
<ipython-input-2-e1d6d02030d3> in <module>
      1 import tensorflow_hub as tfhub
      2 import tensorflow as tf
----> 3 import tensorflow_text

~/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/__init__.py in <module>
     20 from tensorflow_text.python import keras
     21 from tensorflow_text.python import metrics
---> 22 from tensorflow_text.python.ops import *
     23 
     24 

~/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/__init__.py in <module>
     16 """Various TensorFlow ops related to text-processing."""
     17 
---> 18 from tensorflow_text.python.ops.bert_tokenizer import BertTokenizer
     19 from tensorflow_text.python.ops.create_feature_bitmask_op import create_feature_bitmask
     20 from tensorflow_text.python.ops.greedy_constrained_sequence_op import greedy_constrained_sequence

~/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/bert_tokenizer.py in <module>
     28 from tensorflow_text.python.ops.tokenization import Tokenizer
     29 from tensorflow_text.python.ops.tokenization import TokenizerWithOffsets
---> 30 from tensorflow_text.python.ops.wordpiece_tokenizer import WordpieceTokenizer
     31 
     32 

~/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/wordpiece_tokenizer.py in <module>
     32 from tensorflow.python.framework import load_library
     33 from tensorflow.python.platform import resource_loader
---> 34 gen_wordpiece_tokenizer = load_library.load_op_library(resource_loader.get_path_to_datafile('_wordpiece_tokenizer.dylib'))
     35 
     36 

~/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow/python/framework/load_library.py in load_op_library(library_filename)
     56     RuntimeError: when unable to load the library or get the python wrappers.
     57   """
---> 58   lib_handle = py_tf.TF_LoadLibrary(library_filename)
     59   try:
     60     wrappers = _pywrap_python_op_gen.GetPythonWrappers(

NotFoundError: dlopen(/Users/vincent/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/_wordpiece_tokenizer.dylib, 6): Symbol not found: __ZNK10tensorflow11ResourceMgr8DoLookupERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEENS1_10type_indexES9_PPNS_12ResourceBaseE
  Referenced from: /Users/vincent/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/_wordpiece_tokenizer.dylib
  Expected in: /Users/vincent/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow/python/../libtensorflow_framework.2.dylib
 in /Users/vincent/Development/whatlies/venv/lib/python3.7/site-packages/tensorflow_text/python/ops/_wordpiece_tokenizer.dylib
```

**Comment by koaning** [comment]
In order to get ConveRT to work I had to get an ugly fix from [this issue](https://github.com/tensorflow/text/issues/272).

![image](https://user-images.githubusercontent.com/1019791/88400323-7519cd80-cdc8-11ea-93b9-c14bf1996946.png)

Just to give an impression on how tricky dependencies might get.

**Comment by mkaze** [comment]
@koaning Yeah, unfortunately TF is moving too fast in my opinion, and believe me there were (and still is) a lot of weird issues since they started developing 2.x branch and make eager mode the default. But on the bright side those kind of issues, including the one you mentioned which is platform-specific, gets quickly spotted and fixed by TF dev team in the next release. So those instances shouldn't make you to stick to one specific version of TF (which I can assure you neither of 2.x versions is completely stable, as I have worked with all of them), especially for a library that intends to support a variety of models trained with TF 2.x (i.e. if it was a single model the story would be different). So that's why I suggested [here](https://github.com/RasaHQ/whatlies/issues/92#issuecomment-663554053) to stick to TF 2.x, which that configuration would install the latest stable version of TF and its related packages, which are mostly backwards compatible as well. 

Anyways, that was just my two cents and I don't want to take your time with just mere words and discussion. You hold the reins and fortunately have direct access to user feedback as well (which I don't have!); therefore, you know much better how this library should move forward. I'll work on the huggingface and tf-hub integrations and see if I can manage them to integrate nicely.
